Step to click on a button works incorrectly (finds an element, but fails)

**Describe the bug**
When using step **When I click on a button with the name '<button name>'**, it finds an element but fails with exception

java.lang.AssertionError: Could not move to the element because of an error: org.openqa.selenium.interactions.MoveTargetOutOfBoundsException: move target out of bounds
  (Session info: chrome=79.0.3945.130)
Build info: version: '4.0.0-alpha-1', revision: 'd1d3728cae', time: '2019-04-24T13:42:21'
System info: host: 'EPBYMINW0994', ip: '10.6.96.125', os.name: 'Windows 10', os.arch: 'amd64', os.version: '10.0', java.version: '13.0.2'
Driver info: org.openqa.selenium.chrome.ChromeDriver
Capabilities {acceptInsecureCerts: false, browserName: chrome, browserVersion: 79.0.3945.130, chrome: {chromedriverVersion: 79.0.3945.36 (3582db32b3389..., userDataDir: C:\Users\VERANI~1\AppData\L...}, goog:chromeOptions: {debuggerAddress: localhost:57040}, javascriptEnabled: true, networkConnectionEnabled: false, pageLoadStrategy: normal, platform: WINDOWS, platformName: WINDOWS, proxy: Proxy(), setWindowRect: true, strictFileInteractability: false, timeouts: {implicit: 0, pageLoad: 300000, script: 30000}, unhandledPromptBehavior: dismiss and notify}
Session ID: ff30ec2fffba3ad5b81e9ce6aa9e1ec0
	at org.vividus.ui.web.action.MouseActions.moveToElement(MouseActions.java:198)
	at org.vividus.ui.web.action.MouseActions.click(MouseActions.java:74)
	at org.vividus.ui.web.action.MouseActions.click(MouseActions.java:62)
	at org.vividus.bdd.steps.ui.web.ElementSteps.clickOnElement(ElementSteps.java:308)

![image](https://user-images.githubusercontent.com/60783053/74820667-7889c400-5313-11ea-9248-0407f1b150e6.png)

--- a/vividus-plugin-web-app/src/main/java/org/vividus/ui/web/action/IJavascriptActions.java
+++ b/vividus-plugin-web-app/src/main/java/org/vividus/ui/web/action/IJavascriptActions.java
@@ -49,6 +49,8 @@ public interface IJavascriptActions
 
     void scrollIntoView(WebElement webElement, boolean alignedToTheTop);
 
+    void scrollElementIntoViewportCenter(WebElement webElement);
+
     /**
      * Scrolls to the end of the page with dynamically loading content upon scrolling
      */

--- a/vividus-plugin-web-app/src/main/java/org/vividus/ui/web/action/JavascriptActions.java
+++ b/vividus-plugin-web-app/src/main/java/org/vividus/ui/web/action/JavascriptActions.java
@@ -88,6 +88,12 @@ public class JavascriptActions implements IJavascriptActions, IWebApplicationLis
     }
 
     @Override
+    public void scrollElementIntoViewportCenter(WebElement webElement)
+    {
+        executeAsyncScriptFromResource(JavascriptActions.class, "scroll-element-into-viewport-center.js", webElement);
+    }
+
+    @Override
     public void scrollToEndOfPage()
     {
         executeAsyncScriptFromResource(JavascriptActions.class, "scroll-to-end-of-page.js");

--- a/vividus-plugin-web-app/src/main/java/org/vividus/ui/web/action/MouseActions.java
+++ b/vividus-plugin-web-app/src/main/java/org/vividus/ui/web/action/MouseActions.java
@@ -72,7 +72,7 @@ public class MouseActions implements IMouseActions
             WebElement page = webDriver.findElement(BODY_XPATH_LOCATOR);
             try
             {
-                moveToElement(element);
+                javascriptActions.scrollElementIntoViewportCenter(element);
                 element.click();
                 afterClick(clickResult, page, webDriver, defaultAlertAction);
             }

--- a/vividus-plugin-web-app/src/main/resources/org/vividus/ui/web/action/scroll-element-into-viewport-center.js
+++ b/vividus-plugin-web-app/src/main/resources/org/vividus/ui/web/action/scroll-element-into-viewport-center.js
@@ -1,0 +1,13 @@
+var elementToScroll = arguments[0];
+var exit = arguments[arguments.length-1];
+(scrollElementIntoViewportCenter = function() {
+    var currentWindow = window;
+    while (currentWindow !== window.top) {
+        elementToScroll = currentWindow.frameElement;
+        currentWindow = currentWindow.parent;
+    }
+    currentWindow.scrollBy(0, elementToScroll.getBoundingClientRect().top - currentWindow.innerHeight * 0.25);
+    setTimeout(scrollElementIntoViewportCenter, 500);
+    exit();
+}) ();
+

--- a/vividus-plugin-web-app/src/test/java/org/vividus/ui/web/action/JavascriptActionsTests.java
+++ b/vividus-plugin-web-app/src/test/java/org/vividus/ui/web/action/JavascriptActionsTests.java
@@ -17,6 +17,7 @@
 package org.vividus.ui.web.action;
 
 import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.mockito.Mockito.mock;
 import static org.mockito.Mockito.times;
 import static org.mockito.Mockito.verify;
 import static org.mockito.Mockito.when;
@@ -65,9 +66,6 @@ class JavascriptActionsTests
 
     @Mock
     private IWebDriverManager webDriverManager;
-
-    @Mock
-    private WebElement mockedWebElement;
 
     @Mock(extraInterfaces = { JavascriptExecutor.class, HasCapabilities.class })
     private WebDriver webDriver;
@@ -120,9 +118,20 @@ class JavascriptActionsTests
     @Test
     void testScrollIntoView()
     {
-        javascriptActions.scrollIntoView(mockedWebElement, true);
-        verify((JavascriptExecutor) webDriver).executeScript("arguments[0].scrollIntoView(arguments[1])",
-                mockedWebElement, true);
+        WebElement webElement = mock(WebElement.class);
+        javascriptActions.scrollIntoView(webElement, true);
+        verify((JavascriptExecutor) webDriver).executeScript("arguments[0].scrollIntoView(arguments[1])", webElement,
+                true);
+    }
+
+    @Test
+    void testScrollElementIntoViewportCenter()
+    {
+        WebElement webElement = mock(WebElement.class);
+        javascriptActions.scrollElementIntoViewportCenter(webElement);
+        verify((JavascriptExecutor) webDriver).executeAsyncScript(
+                ResourceUtils.loadResource(JavascriptActionsTests.class, "scroll-element-into-viewport-center.js"),
+                webElement);
     }
 
     @Test
@@ -144,16 +153,18 @@ class JavascriptActionsTests
     @Test
     void testScrollToEndOf()
     {
-        javascriptActions.scrollToEndOf(mockedWebElement);
+        WebElement webElement = mock(WebElement.class);
+        javascriptActions.scrollToEndOf(webElement);
         verify((JavascriptExecutor) webDriver).executeScript("arguments[0].scrollTop = arguments[0].scrollHeight",
-                mockedWebElement);
+                webElement);
     }
 
     @Test
     void testScrollToStartOf()
     {
-        javascriptActions.scrollToStartOf(mockedWebElement);
-        verify((JavascriptExecutor) webDriver).executeScript("arguments[0].scrollTop = 0", mockedWebElement);
+        WebElement webElement = mock(WebElement.class);
+        javascriptActions.scrollToStartOf(webElement);
+        verify((JavascriptExecutor) webDriver).executeScript("arguments[0].scrollTop = 0", webElement);
     }
 
     @Test
@@ -167,18 +178,20 @@ class JavascriptActionsTests
     @Test
     void testTriggerEvents()
     {
+        WebElement webElement = mock(WebElement.class);
         String script = "if(document.createEvent){var evObj = document.createEvent('MouseEvents');"
                 + "evObj.initEvent('click', true, false); arguments[0].dispatchEvent(evObj);}"
                 + " else if(document.createEventObject) { arguments[0].fireEvent('onclick');}";
-        javascriptActions.triggerMouseEvents(mockedWebElement, "click");
-        verify((JavascriptExecutor) webDriver).executeScript(script, mockedWebElement);
+        javascriptActions.triggerMouseEvents(webElement, "click");
+        verify((JavascriptExecutor) webDriver).executeScript(script, webElement);
     }
 
     @Test
     void testClick()
     {
-        javascriptActions.click(mockedWebElement);
-        verify((JavascriptExecutor) webDriver).executeScript("arguments[0].click()", mockedWebElement);
+        WebElement webElement = mock(WebElement.class);
+        javascriptActions.click(webElement);
+        verify((JavascriptExecutor) webDriver).executeScript("arguments[0].click()", webElement);
     }
 
     @Test
@@ -201,17 +214,19 @@ class JavascriptActionsTests
     @Test
     void testGetElementTextNotFirefox()
     {
+        WebElement webElement = mock(WebElement.class);
         mockIsFirefox(false);
-        when(((JavascriptExecutor) webDriver).executeScript(ELEMENT_INNER_TEXT, mockedWebElement)).thenReturn(TEXT);
-        assertEquals(TEXT, javascriptActions.getElementText(mockedWebElement));
+        when(((JavascriptExecutor) webDriver).executeScript(ELEMENT_INNER_TEXT, webElement)).thenReturn(TEXT);
+        assertEquals(TEXT, javascriptActions.getElementText(webElement));
     }
 
     @Test
     void testGetElementValue()
     {
-        when(((JavascriptExecutor) webDriver).executeScript("return arguments[0].value", mockedWebElement))
+        WebElement webElement = mock(WebElement.class);
+        when(((JavascriptExecutor) webDriver).executeScript("return arguments[0].value", webElement))
                 .thenReturn(TEXT);
-        assertEquals(TEXT, javascriptActions.getElementValue(mockedWebElement));
+        assertEquals(TEXT, javascriptActions.getElementValue(webElement));
     }
 
     @Test
@@ -235,20 +250,21 @@ class JavascriptActionsTests
     @Test
     void testGetElementAttributesValues()
     {
-        Map<String, String> attributes = new HashMap<>();
-        attributes.put(TEXT, TEXT);
-        when(((JavascriptExecutor) webDriver).executeScript(SCRIPT_GET_ELEMENT_ATTRIBUTES, mockedWebElement))
+        WebElement webElement = mock(WebElement.class);
+        Map<String, String> attributes = Map.of(TEXT, TEXT);
+        when(((JavascriptExecutor) webDriver).executeScript(SCRIPT_GET_ELEMENT_ATTRIBUTES, webElement))
                 .thenReturn(attributes);
-        assertEquals(javascriptActions.getElementAttributes(mockedWebElement), attributes);
+        assertEquals(javascriptActions.getElementAttributes(webElement), attributes);
     }
 
     @Test
     void testSetElementTopPosition()
     {
+        WebElement webElement = mock(WebElement.class);
         Long top = 10L;
-        when(((JavascriptExecutor) webDriver).executeScript(String.format(SCRIPT_SET_TOP_POSITION, top),
-                mockedWebElement)).thenReturn(top);
-        int resultTop = javascriptActions.setElementTopPosition(mockedWebElement, top.intValue());
+        when(((JavascriptExecutor) webDriver).executeScript(String.format(SCRIPT_SET_TOP_POSITION, top), webElement))
+                .thenReturn(top);
+        int resultTop = javascriptActions.setElementTopPosition(webElement, top.intValue());
         assertEquals(top.intValue(), resultTop);
     }
 
@@ -270,16 +286,18 @@ class JavascriptActionsTests
     @Test
     void shouldScrollToTheStartOfWebElement()
     {
-        javascriptActions.scrollToLeftOf(mockedWebElement);
-        verify((JavascriptExecutor) webDriver).executeScript("arguments[0].scrollLeft=0;", mockedWebElement);
+        WebElement webElement = mock(WebElement.class);
+        javascriptActions.scrollToLeftOf(webElement);
+        verify((JavascriptExecutor) webDriver).executeScript("arguments[0].scrollLeft=0;", webElement);
     }
 
     @Test
     void shouldScrollToTheEndOfWebElement()
     {
-        javascriptActions.scrollToRightOf(mockedWebElement);
+        WebElement webElement = mock(WebElement.class);
+        javascriptActions.scrollToRightOf(webElement);
         verify((JavascriptExecutor) webDriver).executeScript("arguments[0].scrollLeft=arguments[0].scrollWidth;",
-                mockedWebElement);
+                webElement);
     }
 
     private void mockScriptExecution(String script, Object result)

--- a/vividus-plugin-web-app/src/test/java/org/vividus/ui/web/action/MouseActionsTests.java
+++ b/vividus-plugin-web-app/src/test/java/org/vividus/ui/web/action/MouseActionsTests.java
@@ -23,12 +23,13 @@ import static org.junit.jupiter.api.Assertions.assertFalse;
 import static org.junit.jupiter.api.Assertions.assertThrows;
 import static org.junit.jupiter.api.Assertions.assertTrue;
 import static org.mockito.ArgumentMatchers.any;
-import static org.mockito.Mockito.doNothing;
 import static org.mockito.Mockito.doReturn;
 import static org.mockito.Mockito.doThrow;
+import static org.mockito.Mockito.inOrder;
 import static org.mockito.Mockito.mock;
 import static org.mockito.Mockito.never;
 import static org.mockito.Mockito.spy;
+import static org.mockito.Mockito.times;
 import static org.mockito.Mockito.verify;
 import static org.mockito.Mockito.verifyNoInteractions;
 import static org.mockito.Mockito.when;
@@ -42,6 +43,7 @@ import com.google.common.eventbus.EventBus;
 import org.junit.jupiter.api.Test;
 import org.junit.jupiter.api.extension.ExtendWith;
 import org.mockito.ArgumentMatchers;
+import org.mockito.InOrder;
 import org.mockito.InjectMocks;
 import org.mockito.Mock;
 import org.mockito.junit.jupiter.MockitoExtension;
@@ -115,37 +117,34 @@ class MouseActionsTests
     @InjectMocks
     private MouseActions mouseActions;
 
-    private void verifyWebElement(MouseActions mouseActionsSpy, boolean alertPresent, boolean newPageLoaded,
-            ClickResult result)
+    private void verifyWebElement(int clickAttempts, boolean alertPresent, boolean newPageLoaded, ClickResult result)
     {
-        verify(webElement).click();
+        verify(webElement, times(clickAttempts)).click();
         assertTrue(result.isClicked());
-        verify(mouseActionsSpy).moveToElement(webElement);
         assertEquals(newPageLoaded, result.isNewPageLoaded());
+        InOrder ordered = inOrder(javascriptActions, alertActions, waitActions, eventBus, webUiContext);
+        ordered.verify(javascriptActions).scrollElementIntoViewportCenter(webElement);
+        ordered.verify(alertActions).isAlertPresent();
         if (!alertPresent)
         {
-            verify(waitActions).waitForPageLoad();
-            verify(eventBus)
+            ordered.verify(waitActions).waitForPageLoad();
+            ordered.verify(alertActions).waitForAlert(webDriver);
+            ordered.verify(eventBus)
                     .post(ArgumentMatchers.<PageLoadEndEvent>argThat(arg -> arg.isNewPageLoaded() == newPageLoaded));
         }
-        else
-        {
-            verifyNoInteractions(eventBus);
-        }
-    }
-
-    private MouseActions mockMoveToElement()
-    {
-        MouseActions mouseActionsSpy = spy(this.mouseActions);
-        doNothing().when(mouseActionsSpy).moveToElement(webElement);
-        return mouseActionsSpy;
+        ordered.verifyNoMoreInteractions();
     }
 
     private void testClick(boolean alertPresent, boolean newPageLoaded)
     {
-        MouseActions mouseActionsSpy = mockMoveToElement();
-        ClickResult result = mouseActionsSpy.click(webElement);
-        verifyWebElement(mouseActionsSpy, alertPresent, newPageLoaded, result);
+        ClickResult result = mouseActions.click(webElement);
+        verifyWebElement(1, alertPresent, newPageLoaded, result);
+    }
+
+    private void testClickWithElementNotClickableException()
+    {
+        ClickResult result = mouseActions.click(webElement);
+        verifyWebElement(2, false, false, result);
     }
 
     @Test
@@ -224,13 +223,13 @@ class MouseActionsTests
         WebDriverException e = new WebDriverException(ELEMENT_IS_NOT_CLICKABLE_AT_POINT);
         WebDriverException e2 = new WebDriverException(STALE_EXCEPTION);
         doThrow(e).doThrow(e2).when(webElement).click();
-        MouseActions mouseActionsSpy = mockMoveToElement();
-        ClickResult result = mouseActionsSpy.click(webElement);
+        ClickResult result = mouseActions.click(webElement);
         verify(webElement, never()).sendKeys("");
         assertFalse(result.isNewPageLoaded());
-        verify(softAssert).recordFailedAssertion(COULD_NOT_CLICK_ERROR_MESSAGE + e2);
-        verify(alertActions, never()).waitForAlert(webDriver);
-        verifyNoInteractions(webUiContext);
+        InOrder ordered = inOrder(javascriptActions, alertActions, eventBus, webUiContext, softAssert);
+        ordered.verify(javascriptActions).scrollElementIntoViewportCenter(webElement);
+        ordered.verify(softAssert).recordFailedAssertion(COULD_NOT_CLICK_ERROR_MESSAGE + e2);
+        ordered.verifyNoMoreInteractions();
     }
 
     @Test
@@ -277,12 +276,8 @@ class MouseActionsTests
         WebDriverException e = new WebDriverException(ELEMENT_IS_NOT_CLICKABLE_AT_POINT);
         doThrow(e).doNothing().when(webElement).click();
         when(alertActions.isAlertPresent()).thenReturn(Boolean.FALSE);
-        MouseActions mouseActionsSpy = mockMoveToElement();
-        ClickResult result = mouseActionsSpy.click(webElement);
+        testClickWithElementNotClickableException();
         verify(webElement, never()).sendKeys("");
-        assertFalse(result.isNewPageLoaded());
-        verify(alertActions).waitForAlert(webDriver);
-        verifyNoInteractions(webUiContext);
     }
 
     @Test
@@ -294,11 +289,11 @@ class MouseActionsTests
         WebDriverException e = new WebDriverException(ELEMENT_IS_NOT_CLICKABLE_AT_POINT);
         WebDriverException e2 = new WebDriverException(STALE_EXCEPTION);
         doThrow(e).doThrow(e2).when(webElement).click();
-        MouseActions mouseActionsSpy = mockMoveToElement();
-        mouseActionsSpy.click(webElement);
-        verify(softAssert).recordFailedAssertion(COULD_NOT_CLICK_ERROR_MESSAGE + e2);
-        verify(alertActions, never()).waitForAlert(webDriver);
-        verifyNoInteractions(webUiContext);
+        mouseActions.click(webElement);
+        InOrder ordered = inOrder(javascriptActions, alertActions, eventBus, webUiContext, softAssert);
+        ordered.verify(javascriptActions).scrollElementIntoViewportCenter(webElement);
+        ordered.verify(softAssert).recordFailedAssertion(COULD_NOT_CLICK_ERROR_MESSAGE + e2);
+        ordered.verifyNoMoreInteractions();
     }
 
     @Test
@@ -309,11 +304,7 @@ class MouseActionsTests
 
         WebDriverException e = new WebDriverException(ELEMENT_IS_NOT_CLICKABLE_AT_POINT);
         doThrow(e).doNothing().when(webElement).click();
-        MouseActions mouseActionsSpy = mockMoveToElement();
-        ClickResult result = mouseActionsSpy.click(webElement);
-        assertFalse(result.isNewPageLoaded());
-        verify(alertActions).waitForAlert(webDriver);
-        verify(eventBus).post(any(PageLoadEndEvent.class));
+        testClickWithElementNotClickableException();
     }
 
     @Test
@@ -324,13 +315,14 @@ class MouseActionsTests
 
         WebDriverException e = new WebDriverException(OTHER_ELEMENT_WOULD_RECEIVE_CLICK);
         doThrow(e).when(webElement).click();
-        MouseActions mouseActionsSpy = mockMoveToElement();
-        ClickResult result = mouseActionsSpy.click(webElement);
-        verify(javascriptActions).click(webElement);
+        ClickResult result = mouseActions.click(webElement);
         assertFalse(result.isNewPageLoaded());
-        verify(alertActions).waitForAlert(webDriver);
-        verify(eventBus).post(any(PageLoadEndEvent.class));
-        verifyNoInteractions(webUiContext);
+        InOrder ordered = inOrder(javascriptActions, alertActions, eventBus, webUiContext, softAssert);
+        ordered.verify(javascriptActions).scrollElementIntoViewportCenter(webElement);
+        ordered.verify(javascriptActions).click(webElement);
+        ordered.verify(alertActions).waitForAlert(webDriver);
+        ordered.verify(eventBus).post(any(PageLoadEndEvent.class));
+        ordered.verifyNoMoreInteractions();
     }
 
     @Test
@@ -341,13 +333,7 @@ class MouseActionsTests
 
         WebDriverException e = new WebDriverException(OTHER_ELEMENT_WOULD_RECEIVE_CLICK);
         doThrow(e).doNothing().when(webElement).click();
-        MouseActions mouseActionsSpy = mockMoveToElement();
-        ClickResult result = mouseActionsSpy.click(webElement);
-        verifyNoInteractions(javascriptActions);
-        assertFalse(result.isNewPageLoaded());
-        verify(alertActions).waitForAlert(webDriver);
-        verify(eventBus).post(any(PageLoadEndEvent.class));
-        verifyNoInteractions(webUiContext);
+        testClickWithElementNotClickableException();
     }
 
     @Test
@@ -358,12 +344,7 @@ class MouseActionsTests
 
         WebDriverException e = new WebDriverException(ELEMENT_IS_NOT_CLICKABLE_AT_POINT);
         doThrow(e).doNothing().when(webElement).click();
-        MouseActions mouseActionsSpy = mockMoveToElement();
-        ClickResult result = mouseActionsSpy.click(webElement);
-        assertFalse(result.isNewPageLoaded());
-        verify(alertActions).waitForAlert(webDriver);
-        verify(eventBus).post(any(PageLoadEndEvent.class));
-        verifyNoInteractions(webUiContext);
+        testClickWithElementNotClickableException();
     }
 
     @Test
@@ -401,10 +382,8 @@ class MouseActionsTests
         boolean alertPresent = true;
         mockBodySearch();
         when(alertActions.isAlertPresent()).thenReturn(alertPresent);
-        MouseActions mouseActionsSpy = mockMoveToElement();
-        ClickResult result = mouseActionsSpy.click(webElement, Optional.of(Action.ACCEPT));
-        verifyWebElement(mouseActionsSpy, alertPresent, false, result);
-        verifyNoInteractions(webUiContext);
+        ClickResult result = mouseActions.click(webElement, Optional.of(Action.ACCEPT));
+        verifyWebElement(1, alertPresent, false, result);
     }
 
     private WebElement mockBodySearch()
@@ -467,9 +446,7 @@ class MouseActionsTests
     void testMoveToNullElement()
     {
         mouseActions.moveToElement(null);
-        verifyNoInteractions(webDriverProvider);
-        verifyNoInteractions(javascriptActions);
-        verifyNoInteractions(softAssert);
+        verifyNoInteractions(webDriverProvider, javascriptActions, softAssert);
     }
 
     @Test

